synapse/util/caches/stream_change_cache.py

```diff
@@ -128,9 +131,11 @@ def entity_has_changed(self, entity, stream_pos):
             self._cache[stream_pos] = entity
             self._entity_to_key[entity] = stream_pos
 
-            while len(self._cache) > self._max_size:
-                k, r = self._cache.popitem()
-                self._earliest_known_stream_pos = max(k, self._earliest_known_stream_pos)
+            while len(self._cache) > self._max_size + 1:
```

Comment:
what's with the + 1 ?

I mean, I see you have tests which assert this is right. I just can't see how :)

Reply:
That was a good question, and then I realised that max_size was * the cache factor.

Now, it mocks out the cache factor, and this mis-solution is removed :)